src/psyclone/psyGen.py

```diff
@@ -2542,14 +2574,31 @@ def update(self):
         enddir = Comment(FortranStringReader("!$omp end parallel",
                                              ignore_comments=False))
         self.ast_end = enddir
-        # If end_idx+1 takes us beyond the range of the list then the
-        # element is appended to the list
-        self._parent.ast.content.insert(end_idx+1, enddir)
+        # FIXME: In case of:
+        # omp parallel
+        #    omp do
+        #       loop
+        #    omp end do
+        # omp end parallel
+        # the "end parallel" statement will be inserted after the "omp do"
+        # statement!
+        # In order to avoid this problem when an "omp do" is present, test
+        # for this case and if so move the "omp end parallel" two statements
+        # further down, i.e. after the loop and "omp end do" statement.
+        if isinstance(parent_ast.content[end_idx], Comment) and \
+                "omp do" in str(parent_ast.content[end_idx]):
+            # We need to test for instance, otherwise the string representation
+            # of a loop could somewhere contain an "omp do"
+            parent_ast.content.insert(end_idx+3, enddir)
```

Comment:
Please could you add a check here that `end_idx+3` really is the right location - i.e. that the nodes either side are what you expect. That way, if this 'fix' fails then at least we'll exit cleanly.

Reply:
Hmm - I did add one more test, but tbh there is only so much we can test. The test for end_idx+2 only confirms that an omp-do has exactly one inner node (the loop), and one end-do. The omp do transform should already guarantee that.